### PR TITLE
camel-fury: Improve to allow auto discovery Fury from the registry

### PR DESCRIPTION
**Pull request by zhfeng**
* replace to use BaseFury
* add an option with allowAutoWiredFury
* add FuryDataFormatReifier

# Description

# Target

- [ ] I checked that the commit is targeting the correct branch (note that Camel 3 uses `camel-3.x`, whereas Camel 4 uses the `main` branch)

# Tracking
- [ ] If this is a large change, bug fix, or code improvement, I checked there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).

# Apache Camel coding standards and style

- [ ] I checked that each commit in the pull request has a meaningful subject line and body.

- [ ] I have run `mvn clean install -DskipTests` locally from root folder and I have committed all auto-generated changes.

### CODE REVIEW

File: core/camel-core-model/src/main/java/org/apache/camel/model/dataformat/FuryDataFormat.java [hunk 1]
```diff
@@ -40,6 +40,19 @@ public class FuryDataFormat extends DataFormatDefinition {
     @Metadata(description = "Class of the java type to use when unmarshalling")
     private String unmarshalTypeName;
 
+    @XmlAttribute
+    @Metadata(description = "Whether to require register classes", defaultValue = "True", javaType = "java.lang.Boolean")
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/dataformat/FuryDataFormat.java:44]: true should be lower case

File: components/camel-fury/src/main/java/org/apache/camel/component/fury/FuryDataFormat.java [hunk 2]
```diff
@@ -85,7 +91,21 @@ protected void doInit() throws Exception {
             unmarshalType = camelContext.getClassResolver().resolveClass(unmarshalTypeName);
         }
 
-        fury = Fury.builder().withLanguage(Language.JAVA).requireClassRegistration(true).build();
+        if (fury == null && isAllowAutoWiredFury()) {
+            Set<BaseFury> beans = getCamelContext().getRegistry().findByType(BaseFury.class);
```
Comment (davsclaus) [inline components/camel-fury/src/main/java/org/apache/camel/component/fury/FuryDataFormat.java:95]: Use `findSingleByType` instead

File: core/camel-core-model/src/main/java/org/apache/camel/model/dataformat/FuryDataFormat.java [hunk 3]
```diff
@@ -40,6 +40,19 @@ public class FuryDataFormat extends DataFormatDefinition {
     @Metadata(description = "Class of the java type to use when unmarshalling")
     private String unmarshalTypeName;
 
+    @XmlAttribute
+    @Metadata(description = "Whether to require register classes", defaultValue = "true", javaType = "java.lang.Boolean")
+    private String requireClassRegistration;
+
+    @XmlAttribute
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/dataformat/FuryDataFormat.java:47]: this thread safe and autowired options could be labelled as advanced

File: core/camel-core-model/src/main/java/org/apache/camel/model/MarshalDefinition.java [hunk 4]
```diff
@@ -88,6 +89,7 @@ public class MarshalDefinition extends NoOutputDefinition<MarshalDefinition> imp
             @XmlElement(name = "custom", type = CustomDataFormat.class),
             @XmlElement(name = "fhirJson", type = FhirJsonDataFormat.class),
             @XmlElement(name = "fhirXml", type = FhirXmlDataFormat.class),
+            @XmlElement(name = "fury", type = FuryDataFormat.class),
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/MarshalDefinition.java:92]: You should add this in other places, just grep for one of the existing ie in Unmarshal and DataFormats definition classes

**Review by davsclaus — CHANGES_REQUESTED** [review]

File: core/camel-core-model/src/main/java/org/apache/camel/model/MarshalDefinition.java (hunk 4 above)
Comment (zhfeng) [reply core/camel-core-model/src/main/java/org/apache/camel/model/MarshalDefinition.java]: Thanks @davsclaus for pointing out. Just wonder if there is any docs for develop a new `DataFormat`?

Comment (davsclaus) [reply core/camel-core-model/src/main/java/org/apache/camel/model/MarshalDefinition.java]: No specific docs - the best is looking at any of the others and see how they are done.

**Review by davsclaus — APPROVED** [review]

**Pull request merged**